provider/infoblox.go

```diff
@@ -0,0 +1,483 @@
+
+		var resT []ibclient.RecordTXT
+		objT := ibclient.NewRecordTXT(
+			ibclient.RecordTXT{
+				Zone: zone.Fqdn,
+			},
+		)
+		err = p.client.GetObject(objT, "", &resT)
+		if err != nil {
+			return nil, err
+		}
+		for _, res := range resT {
+			// The Infoblox API strips enclosing double quotes from TXT records lacking whitespace.
+			// Unhandled, the missing double quotes would break the extractOwnerID method of the registry package.
+			if _, err := strconv.Unquote(res.Text); err != nil {
+				res.Text = strconv.Quote(res.Text)
+			}
+			endpoints = append(endpoints, endpoint.NewEndpoint(res.Name, res.Text, endpoint.RecordTypeTXT))
+		}
+	}
+	return endpoints, nil
+}
+
+// ApplyChanges applies the given changes.
+func (p *InfobloxProvider) ApplyChanges(changes *plan.Changes) error {
+	zones, err := p.zones()
+	if err != nil {
+		return err
+	}
+
+	created, deleted, updated := p.mapChanges(zones, changes)
+	p.createRecords(created)
+	p.deleteRecords(deleted)
+	p.updateRecords(updated)
+	return nil
+}
+
+func (p *InfobloxProvider) zones() ([]ibclient.ZoneAuth, error) {
+	var res, result []ibclient.ZoneAuth
+	obj := ibclient.NewZoneAuth(ibclient.ZoneAuth{})
+	err := p.client.GetObject(obj, "", &res)
+
+	if err != nil {
+		return nil, err
+	}
+
+	for _, zone := range res {
+		if p.domainFilter.Match(zone.Fqdn) {
+			result = append(result, zone)
+		}
+	}
+
+	return result, nil
+}
+
+type infobloxChangeMap map[string][]*endpoint.Endpoint
+
+func (p *InfobloxProvider) mapChanges(zones []ibclient.ZoneAuth, changes *plan.Changes) (infobloxChangeMap, infobloxChangeMap, infobloxChangeMap) {
+	created := infobloxChangeMap{}
+	deleted := infobloxChangeMap{}
+	updated := infobloxChangeMap{}
+
+	mapChange := func(changeMap infobloxChangeMap, change *endpoint.Endpoint) {
+		zone := p.findZone(zones, change.DNSName)
+		if zone == nil {
+			logrus.Infof("Ignoring changes to '%s' because a suitable Infoblox DNS zone was not found.", change.DNSName)
+			return
+		}
+		// Ensure the record type is suitable
+		changeMap[zone.Fqdn] = append(changeMap[zone.Fqdn], change)
+	}
+
+	for _, change := range changes.Delete {
+		mapChange(deleted, change)
+	}
+	for _, change := range changes.UpdateOld {
+		mapChange(deleted, change)
+	}
+	for _, change := range changes.Create {
+		mapChange(created, change)
+	}
+	for _, change := range changes.UpdateNew {
+		mapChange(created, change)
+	}
+
+	return created, deleted, updated
+}
+
+func (p *InfobloxProvider) findZone(zones []ibclient.ZoneAuth, name string) *ibclient.ZoneAuth {
+	var result *ibclient.ZoneAuth
+
+	// Go through every zone looking for the longest name (i.e. most specific) as a matching suffix
+	for idx := range zones {
+		zone := &zones[idx]
+		if strings.HasSuffix(name, zone.Fqdn) {
+			if result == nil || len(zone.Fqdn) > len(result.Fqdn) {
+				result = zone
+			}
+		}
+	}
+	return result
+}
+
+func (p *InfobloxProvider) mapObjRes(ep *endpoint.Endpoint, getObject bool) (objRes ibObjRes, err error) {
```

Comment:
This method I would also consider as code smell. From the name is not clear what it does and it returns something magic.
A newIbObjRes() method (please replace the name!!) could hide a lot of code duplications here:

		if getObject {
			err = p.client.GetObject(obj, "", &res)
			if err != nil {
				return
			}
		}
		objRes = ibObjRes{
			obj: obj,
			res: &res,
		}

Reply:
@szuecs I had actually tried `newIbObjRes()` but floundered with the type conversion. I just tried this again with a shiny new name with similar results.

If the signature is:

```
func (p *InfobloxProvider) newInfobloxRecordSet(obj ibclient.IBObject, res interface{}, getObject bool) (recordSet infobloxRecordSet) {}
```

and I try to invoke it like so (and similarly for the other types)...

```
var res []ibclient.RecordA
recordSet = p.newInfobloxRecordSet(obj, res, getObject)
```

Or even...

```
var res interface{}
recordSet = p.newInfobloxRecordSet(obj, *res.(*[]ibclient.RecordA), getObject)
```

I get the same interface conversion problems. Any pointers for a man who has been ruined by dynamic typing?

Reply:
I spent the better part of the afternoon trying to solve the conundrum I mentioned earlier. I would really like to know the right way to handle this, but I'm begging for a little mercy here so I don't spend another whole day trying to resolve this.